mindmeld/models/model.py

```diff
@@ -1223,104 +654,54 @@ def initialize_resources(self, resource_loader, examples=None, labels=None):
         self._resources["tokenizer"] = resource_loader.get_tokenizer()
 
 
-class LabelEncoder:
-    """The label encoder is responsible for converting between rich label
-    objects such as a ProcessedQuery and basic formats a model can interpret.
-
-    A MindMeld model uses its label encoder at fit time to encode labels into a
-    form it can deal with, and at predict time to decode predictions into
-    objects
-    """
+class PytorchModel(BaseModel):
 
     def __init__(self, config):
-        """Initializes an encoder
-
-        Args:
-            config (ModelConfig): The model
-        """
-        self.config = config
-
-    @staticmethod
-    def encode(labels, **kwargs):
-        """Transforms a list of label objects into a vector of classes.
-
-
-        Args:
-            labels (list): A list of labels to encode
-        """
-        del kwargs
-        return labels
-
-    @staticmethod
-    def decode(classes, **kwargs):
-        """Decodes a vector of classes into a list of labels
-
-        Args:
-            classes (list): A list of classes
-
-        Returns:
-            list: The decoded labels
-        """
-        del kwargs
-        return classes
+        super().__init__(config)
+        config = self.config
+        # self._featurizer = AutoFeaturizer.from_config(config)
 
+    def initialize_resources(self, resource_loader, examples=None, labels=None):
+        # self._featurizer.initialize_resources(examples, labels, params)
+        raise NotImplementedError
 
-class EntityLabelEncoder(LabelEncoder):
-    def __init__(self, config):
-        """Initializes an encoder
+    def fit(self, examples, labels, params=None):
+        # self._featurizer.train()  # this loads an optimizer, batches inputs,
+        # finds loss and backwards gradients and updates gradients and saves best checkpoint
+        raise NotImplementedError
 
-        Args:
-            config (ModelConfig): The model configuration
-        """
-        self.config = config
-        self.system_entity_recognizer = SystemEntityRecognizer.get_instance()
+    def dump(self, path, metadata=None):
+        # do featurizer specific dumping
 
-    def _get_tag_scheme(self):
-        return self.config.model_settings.get("tag_scheme", "IOB").upper()
+        # dump model configs
+        # metadata = metadata or {}
+        # metadata.update({'model_config': self.featurizer.config})
+        super().dump(path)
```

Comment:
Why are we calling this if we are expecting the specific model classes to implement `dump()`?

Reply:
So all models are supposed to dump `model_config` as it is required to determine the right class to instantiate when loading. So calling super().dump(..) validates that and does the dumping of metadata.

Reply:
Once the specific model overrides this method, doesn't it become the responsible of that version of `dump()` to write the model config? This method will just raise a `NotImplementedError`

Reply:
I modified this code snippet in the latest version of code. So the subclasses get to define a method called `_dump()` which is in-turn called in a publicly exposed method `dump()` residing in the base class.